statistics/handle/ddl.go

```diff
@@ -44,21 +46,91 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
 			}
 		}
 	case model.ActionAddTablePartition, model.ActionTruncateTablePartition:
-		pruneMode := h.CurrentPruneMode()
-		if pruneMode == variable.Static {
-			for _, def := range t.PartInfo.Definitions {
-				if err := h.insertTableStats2KV(t.TableInfo, def.ID); err != nil {
-					return err
-				}
+		for _, def := range t.PartInfo.Definitions {
+			if err := h.insertTableStats2KV(t.TableInfo, def.ID); err != nil {
+				return err
 			}
 		}
-		if pruneMode == variable.Dynamic {
-			// TODO: need trigger full analyze
-		}
 	case model.ActionDropTablePartition:
 		pruneMode := h.CurrentPruneMode()
-		if pruneMode == variable.Dynamic {
-			// TODO: need trigger full analyze
+		if pruneMode == variable.Dynamic && t.PartInfo != nil {
+			if err := h.updateGlobalStats(t.TableInfo); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// analyzeOptionDefault saves the default values of NumBuckets and NumTopN.
+// These values will be used in dynamic mode when we drop table partition and then need to merge global-stats.
+// These values originally came from the analyzeOptionDefault structure in the planner/core/planbuilder.go file.
+var analyzeOptionDefault = map[ast.AnalyzeOptionType]uint64{
+	ast.AnalyzeOptNumBuckets: 256,
+	ast.AnalyzeOptNumTopN:    20,
+}
+
+// updateGlobalStats will trigger the merge of global-stats when we drop table partition
+func (h *Handle) updateGlobalStats(tblInfo *model.TableInfo) error {
+	// We need to merge the partition-level stats to global-stats when we drop table partition in dynamic mode.
+	tableID := tblInfo.ID
+	is := infoschema.GetInfoSchema(h.mu.ctx)
+	globalStats, err := h.TableStatsFromStorage(tblInfo, tableID, true, 0)
+	if err != nil {
+		return err
+	}
+	// If we do not currently have global-stats, no new global-stats will be generated.
+	if globalStats != nil {
```

Comment:
`if globalStats == nil {return nil}` is more go-style than `if globalStats != nil {...}`.